### PR TITLE
Ed/typescript conversion

### PR DESCRIPTION
**Pull request by eropple**
Typescript conversion is done

- `types/*` directory is gone
- `tsc` builds in place; `main` is now `dist/index.js`
- CI actions updated to build before doing eslint/test
- Format clean
- Tests clean

**BREAKING CHANGE:**
TypeScript wants you to do `export const blah` or `export class blah` rather than `module.exports = Blah`, and this means that it creates ESModule-style shims. I have _kept_ the CJS export at the root of the repository, but all others are now ESM-style. This means that: `const Mux = require('@mux/node`)` _will still work_, but `const Assets = require('@mux/node/video/resources/assets')` will _not_ (it'll be `const { Assets }`).

As we will want to move towards an ESM module pretty soon (felt out of scope here, and we have other features/rehab work to do here too), I decided to defer the full ESM conversion for another major rev.

### CODE REVIEW

**Comment by eropple** [comment]
One thing I know is wrong but have not had time to look at is that all video stuff was reliant on `ListParams`, which had somehow become a sum type of all video `List` params (so `live_stream_id` was being allowed for the wrong lists, etc.).

I only copied those out, I did not figure out which ones should/shouldn't have it, I would appreciate some help there.

**Comment by eropple** [comment]
(also, eslint is super broken, help there would be awesome or I'll get to it when I can)

**Comment by davekiss** [comment]
Instead, I made the diff super confusing with prettier stuff to please eslint

File: src/base.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      // we could do this with Object.assign but I'd rather we be really explicit about what we copy.
+      this.config = tokenIdOrOptionsOrBase._config;
+      this._tokenId = tokenIdOrOptionsOrBase._tokenId;
+      this._tokenSecret = tokenIdOrOptionsOrBase._tokenSecret;
+
+      this.http = tokenIdOrOptionsOrBase.http;
+    } else {
+      if (
+        typeof tokenIdOrOptionsOrBase === 'object' &&
+        !(tokenIdOrOptionsOrBase instanceof Base)
```
Comment (JacquiManzi) [inline src/base.ts:48]: This is just being overly safe, right? It shouldn't actually make it here given the first evaluation

File: src/base.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      ...options,
+    };
+  }
+
+  get config() {
+    return this._config;
+  }
+
+  set tokenId(token: string | undefined) {
+    const v = token || process.env.MUX_TOKEN_ID;
```
Comment (JacquiManzi) [inline src/base.ts:114]: Nit on using a single letter `v` here, especially since it's being also used below for `tokenSecret`

File: src/base.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          Promise.reject(
+            (errorRes.response && errorRes.response.data.error) || errorRes
+          )
+      );
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  isVideoUrl(url: string) {
+    return url.startsWith(`/video/v1/`);
```
Comment (JacquiManzi) [inline src/base.ts:99]: Nit on back ticks being used without needing templating

File: src/base.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    this._tokenId = v;
+  }
+
+  get tokenId() {
+    return this._tokenId;
+  }
+
+  set tokenSecret(secret: string | undefined) {
+    const v = secret || process.env.MUX_TOKEN_SECRET;
+    if (!v || v.length === 0) {
```
Comment (JacquiManzi) [inline src/base.ts:128]: potentially explore `!!` here

File: src/video/video.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  constructor(accessToken: string, secret: string, config: RequestOptions);
+  constructor(
+    accessTokenOrConfigOrBase: string | RequestOptions | Base,
+    secret?: string,
+    config?: RequestOptions
+  ) {
+    if (accessTokenOrConfigOrBase instanceof Base) {
+      super(accessTokenOrConfigOrBase);
+    } else if (typeof accessTokenOrConfigOrBase === 'object') {
+      super(accessTokenOrConfigOrBase);
```
Comment (JacquiManzi) [inline src/video/video.ts:50]: this check and the above seem like they can be bundled

**Comment by JacquiManzi** [comment]
Looking good! Might be worth exploring TS strict mode

File: src/base.ts (hunk 1 above)
Comment (eropple) [reply src/base.ts:48]: Correct - this is a firewall check

File: src/base.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  constructor(tokenId: string, tokenSecret: string, config: RequestOptions);
+  constructor(
+    tokenIdOrOptionsOrBase: string | RequestOptions | Base,
+    tokenSecret?: string,
+    config?: RequestOptions
+  ) {
+    super();
+
+    if (tokenIdOrOptionsOrBase instanceof Base) {
+      // we could do this with Object.assign but I'd rather we be really explicit about what we copy.
```
[the remark this answers is not in the document]
Comment (eropple) [reply src/base.ts:39]: I was planning on leaving it? "Be human", and also protect ourselves from ourselves in the future. 🤣

File: src/video/resources/assets.ts [hunk 7]
```diff
@@ -61,7 +73,7 @@ class Assets extends Base {
    *
    * @see hhttps://docs.mux.com/api-reference/video#operation/delete-asset
    */
-  del(assetId) {
+  del(assetId: string): Promise<any> {
```
[the remark this answers is not in the document]
Comment (eropple) [reply src/video/resources/assets.ts:76]: This is an import of the existing types. I don't even know what we actually return here; the OAS doc says we return a 204 No Content, but what could actually come out is up in the air.

I guess we could return `unknown`. I'll think about that for a follow-up.

File: src/video/video.ts (hunk 5 above)
Comment (eropple) [reply src/video/video.ts:50]: So when I tried, TypeScript yelled at me. I'll give it another squint before I merge just to be safe.

**Pull request merged**